### ISSUE TITLE
VCH upgrades in a system test setup did not complete and lead to too many open HTTP sessions

### ISSUE THREAD

**Issue by mdubya66** [vmware/vic#7907]
@hchundu-vmw commented on [Tue May 01 2018](https://github.com/vmware/vic-product/issues/1676)

REQUIRED:

vCenter Server version:6.5 U1

Embedded or external PSC:Embedded

Filename of the OVA you deployed:vic-v1.4.0-rc2-4741-f02b3258.ova

How was the OVA deployed? (Flex client, HTML5 client, ovftool): ovftool

Does the VIC appliance recieve configuration by DHCP? Static

What stage of the Appliance Lifecycle is the VIC appliance in? VIC Appliance upgraded succesfully and the VCH upgrade failed

OPTIONAL, but helpful:

IP address of VIC appliance: 10.197.37.183 (old vic appliance on 1.3.1 is 10.197.37.132)
Hostname of VIC appliance: vic-st-h2-130.eng.vmware.com
IP address of vCenter Server: 10.133.146.15
Hostname of vCenter Server: wdc-stvc3944-146-dhcp15.eng.vmware.com 
VCH VM's : 10.197.37.133-10.197.37.181(All the 49 ips in this range are VCH ips)

Bug Report Detailed Information

Vic appliance upgrade succeeded but VCH upgrades did not go through when a total of 50 upgrades were initiated in parallel

Following is the procedure followed

1)Deployed 1.3.1 vic appliance

2)Created 50 VCH's and 2600 containers

3)Created 50 projects

4)Created a network from admiral ui and couple of templates

5)Added all VCH's to admiral

6)Upgraded VIC appliance

7)Tried performing VCH upgrades in parallel ,and only very few of them reported success while all others did not upgrade even after couple of hours ,hence i have aborted the script and tried to verify any problems from Vcenter

The vcenter vpxd reported an error as 

2018-05-01T14:53:52.927Z error vpxd[7FE71D332700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.037Z error vpxd[7FE7F5830700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.089Z error vpxd[7FE724A1E700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.125Z error vpxd[7FE7270E7700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.137Z error vpxd[7FE720C23700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.323Z error vpxd[7FE71F5F7700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.328Z error vpxd[7FE734183700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000
2018-05-01T14:53:53.580Z error vpxd[7FE72B3E8700] [Originator@6876 sub=HTTP session map] Out of HTTP sessions: Limited to 2000

Because of this error i have tried restarting vpxd ,but even after this the vcenter server was reporting too many HTTP sessions

In the vpxd-profiler.log we can see the below trace ,which shows that it is the VCH's which are having these sessions 

--> /SessionStats/SessionPool/Session/Id='52c73db3-ee4f-bd88-df33-6425fe8378e5'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.152'/EventCollector/NumEvents/total 200
--> /SessionStats/SessionPool/Session/Id='52eb30e8-c150-b74a-7701-2387344e9407'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.154'/EventCollector/NumEvents/total 200
--> /SessionStats/SessionPool/Session/Id='52c7f0be-0151-7539-0d48-9e279611a7b7'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.141'/EventCollector/NumEvents/total 200
--> /SessionStats/SessionPool/Session/Id='52338351-1524-4161-a94b-0e2c15a46934'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.137'/EventCollector/NumEvents/total 200
--> /SessionStats/SessionPool/Session/Id='52c7ff8f-d1ee-fbe4-a413-88d726671c8d'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.166'/EventCollector/NumEvents/total 200
--> /SessionStats/SessionPool/Session/Id='521f0b97-c004-e92b-71b4-33524620c6fa'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.177'/EventCollector/NumEvents/total 200
--> /SessionStats/SessionPool/Session/Id='5215fb10-5cd6-a477-2fcc-f9e674377865'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.164'/EventCollector/NumEvents/total 200

--> /SessionStats/SessionPool/Session/Id='52668eb6-29e1-5c07-93f3-e4e951d5102c'/Username='VSPHERE.LOCAL\Administrator'/ClientIP='10.197.37.175'/SessionView/Container/total 51

Even after 2 days that the upgrade was tried ,the out of Http sessions is still seen 

From an environment perspective ,one of the hosts was in not responding state where there were about 5 to 6 VCH's ,which should have failed anyways but all others were up 

From vic-machine.log ,we have errors like below for some of the VCH's and it looks like the reason was that the snapshot creation failed for these VCH's 

Apr 29 2018 02:17:58.402-04:00 ERROR op=11441.1: unexpected fault on task retry: &types.GenericVmConfigFault{VmConfigFault:types.VmConfigFault{VimFault:types.VimFault{MethodFault:types.MethodFault{FaultCause:(*types.LocalizedMethodFault)(nil), FaultMessage:[]types.LocalizableMessage{types.LocalizableMessage{DynamicData:types.DynamicData{}, Key:"msg.vigor.vmx.write.error", Arg:[]types.KeyAnyValue{types.KeyAnyValue{DynamicData:types.DynamicData{}, Key:"1", Value:"/vmfs/volumes/7ff59ee7-827c2183/U-VIC-00044_1/U-VIC-00044.vmx"}}, Message:"Unable to write VMX file: /vmfs/volumes/7ff59ee7-827c2183/U-VIC-00044_1/U-VIC-00044.vmx."}, types.LocalizableMessage{DynamicData:types.DynamicData{}, Key:"msg.fileio.generic", Arg:[]types.KeyAnyValue(nil), Message:"Error"}}}}}, Reason:"Unable to write VMX file: /vmfs/volumes/7ff59ee7-827c2183/U-VIC-00044_1/U-VIC-00044.vmx."}

The live setup is available for debugging ,Please access the vcsa mentioned above because the vc-support bundle is too large to be uploaded here  

---

@hchundu-vmw commented on [Tue May 01 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-385701583)

The vic-machine.log is uploaded to the Vcenter server wdc-stvc3944-146-dhcp15.eng.vmware.com (root/Admin!23) under /storage/imagebuilder ,Also the other vcenter logs can be accessed from this live Vcenter setup to check why the HTTP sessions count got exceeded.

---

@gigawhitlocks commented on [Tue May 01 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-385762138)

Looks like one of the main components (portlayer or personality most likely) is leaking sessions

```
❯ govc session.ls | awk '{print $NF}' | sort | uniq -c                                                                                                                                              
     10 1.0
      1 Agent
      1 cl/1.0.0
     45 Go-http-client/1.1
      1 govc/0.12.1
      1 govc/0.15.0
      9 vic-engine/1.3.1
   1833 vic-engine/1.4.0-rc2

---

@hchundu-vmw commented on [Fri May 04 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386510436)


[sessions_list.txt](https://github.com/vmware/vic-product/files/1973349/sessions_list.txt)


Ran the govc sessions.ls when there were more number of sessions ,This output was taken after a vpxd restart hence the sessions count is 1400 when i took the output ,but eventually it reached 2000 after the output has been captured ,but since these sessions are sufficiently high in number ,I hope this helps 

---

@hchundu-vmw commented on [Fri May 04 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386511607)

Vcenter logs ,vic-machine.log,vic appliance logs can be downloaded from the below location

http://engweb.eng.vmware.com/bugs/files/0/1/8/2/1/3/4/9/vic_1676/

---

@hchundu-vmw commented on [Fri May 04 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386705809)

I have tried recreating this scenario by doing a upgrade from 1.3.1 to 1.4 Rc4 and was able to recreate the scenario

Here are some of the observations

Initially before the upgrade was initiated ,the session count is as below

root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# govc session.ls | wc -l
163

The sessions count started increasing rapidly and after 5 mins ,the count is as below

root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# date
Fri May  4 18:19:23 UTC 2018
root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# govc session.ls | wc -l
752

root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# date
Fri May  4 18:29:16 UTC 2018
root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# govc session.ls | wc -l
1401

In a total time of about 15 mins ,the sessions count reached max of 2000

root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# govc session.ls | wc -l
2008
root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# date
Fri May  4 18:44:25 UTC 2018

After reaaching the maximum ,for some time it looked as if the sessions count started to decrease as below

root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# date
Fri May  4 18:46:49 UTC 2018
root@wdc-stvc3944-146-dhcp15 [ /storage/imagebuilder ]# govc session.ls | wc -l
1904

However it again reached maximum and it stayed there at the maximum number

I will attach the netstat output and also the govc sessions output which shows the details from where the sessions are coming from ,Also the livesetup is available for debugging ,The setup is the same as mentioned in the defect description

---

@hchundu-vmw commented on [Fri May 04 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386706570)

[sessions_rc4_upgrade_2008.txt](https://github.com/vmware/vic-product/files/1975646/sessions_rc4_upgrade_2008.txt)
[netstat_rc4upgrade_1993.txt](https://github.com/vmware/vic-product/files/1975648/netstat_rc4upgrade_1993.txt)


---

@hickeng commented on [Fri May 04 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386771066)

Having dumped [that sessions list](https://github.com/vmware/vic-product/issues/1676#issuecomment-386510436) into a spreadsheet it looks like we have 46 unique sources of sessions, and of those 42 are probably test VCHs (from the 10.197.37.0 subnet).

I've not broken out what the range is for number of sessions from each source, but a casual check suggests it's around 36 sessions. Of the ~1350 VCH related sessions only ~130 have been idle for less than a minute which suggests the others are connections that have been dropped for some reason.


---

@hickeng commented on [Fri May 04 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386776547)

I have not been able to recreate the same session behaviour against a local ESX (6.7) when performing the same upgrade path. The VC in question is non-response at this time.

@hchundu-vmw Please can you grab the VC support bundle (I've slacked you the URL that's responsive) to pass to the vSphere team.
Secondly, could you try (after having recovered VC and cleaned any idle sessions via `govc session.rm`):
1. upgrading a single VCH in that VC env to see how many sessions it generates
2. if (1) doesn't show the multiple vic-engine sessions then try upgrading only 5 in parallel

I'm after a reduced recreation if possible. To grab log bundles from the VCHs, enable debug on one of them before VC dies, then we can use SSH to grab the logs.

---

@hickeng commented on [Sun May 06 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-386948775)

The port layer log shows the following (as a full cycle from 503 to coming up, and back to 503):
```
time="2018-05-04T19:03:50Z" level=info msg="Launching portlayer server pprof server on 127.0.0.1:6063" 
time="2018-05-04T19:03:50Z" level=info msg="{Formatter:0xc420269380 Level:info Syslog:<nil>}" 
May  4 2018 19:03:50.727Z FATAL configure_port_layer ERROR: Failed to connect to wdc-stvc3944-146-dhcp15.eng.vmware.com: 503 Service Unavailable
time="2018-05-04T19:03:53Z" level=info msg="Launching portlayer server pprof server on 127.0.0.1:6063" 
time="2018-05-04T19:03:54Z" level=info msg="{Formatter:0xc420686500 Level:info Syslog:<nil>}" 
May  4 2018 19:03:56.130Z INFO  op=1603.4: Creating directory [CPBU_2TB_01] U-VIC-00023_1/kvStores
May  4 2018 19:03:56.163Z INFO  op=1603.4: Datastore path is [CPBU_2TB_01] U-VIC-00023_1/kvStores
May  4 2018 19:03:56.623Z INFO  NewKeyValueStore(apiKV) restored 3 keys
May  4 2018 19:03:56.623Z INFO  Beginning initialization of portlayer exec component
May  4 2018 19:03:56.754Z INFO  Syncing container cache
May  4 2018 19:03:57.149Z WARN  skipping converting container VM VirtualMachine:vm-142663: could not parse id
May  4 2018 19:03:58.028Z INFO  op=1603.6: Creating directory [CPBU_2TB_01] U-VIC-00023_1/kvStores
May  4 2018 19:03:58.054Z INFO  op=1603.6: Datastore path is [CPBU_2TB_01] U-VIC-00023_1/kvStores
May  4 2018 19:03:58.101Z INFO  NewKeyValueStore(network.contexts.default) restored 0 keys
May  4 2018 19:03:58.101Z INFO  Reserving IP range [172.16.0.0, 172.16.255.255]
May  4 2018 19:03:58.101Z INFO  Reserving IP range [172.16.255.255, 172.16.255.255]
May  4 2018 19:03:58.101Z INFO  Reserving IP range [172.16.0.0, 172.16.0.0]
May  4 2018 19:03:58.101Z INFO  Reserving IP range [172.16.0.1, 172.16.0.1]
May  4 2018 19:03:58.134Z INFO  Reserving IP range [172.16.0.6, 172.16.0.6]
May  4 2018 19:03:58.155Z INFO  Reserving IP range [172.16.0.16, 172.16.0.16]
May  4 2018 19:03:58.179Z INFO  Reserving IP range [172.16.0.38, 172.16.0.38]
May  4 2018 19:03:58.198Z INFO  Reserving IP range [172.16.0.14, 172.16.0.14]
May  4 2018 19:03:58.215Z INFO  Reserving IP range [172.16.0.24, 172.16.0.24]
May  4 2018 19:03:58.252Z INFO  Reserving IP range [172.16.0.25, 172.16.0.25]
May  4 2018 19:03:58.270Z INFO  Reserving IP range [172.16.0.12, 172.16.0.12]
May  4 2018 19:03:58.291Z INFO  Reserving IP range [172.16.0.33, 172.16.0.33]
May  4 2018 19:03:58.307Z INFO  Reserving IP range [172.16.0.34, 172.16.0.34]
May  4 2018 19:03:58.339Z INFO  Reserving IP range [172.16.0.35, 172.16.0.35]
May  4 2018 19:03:58.363Z INFO  Reserving IP range [172.16.0.3, 172.16.0.3]
May  4 2018 19:03:58.382Z INFO  Reserving IP range [172.16.0.5, 172.16.0.5]
May  4 2018 19:03:58.410Z INFO  Reserving IP range [172.16.0.31, 172.16.0.31]
May  4 2018 19:03:58.433Z INFO  Reserving IP range [172.16.0.18, 172.16.0.18]
May  4 2018 19:03:58.460Z INFO  Reserving IP range [172.16.0.8, 172.16.0.8]
May  4 2018 19:03:58.518Z INFO  Reserving IP range [172.16.0.11, 172.16.0.11]
May  4 2018 19:03:58.563Z INFO  Reserving IP range [172.16.0.23, 172.16.0.23]
May  4 2018 19:03:58.596Z INFO  Reserving IP range [172.16.0.30, 172.16.0.30]
May  4 2018 19:03:58.732Z INFO  Reserving IP range [172.16.0.36, 172.16.0.36]
May  4 2018 19:03:58.829Z INFO  Reserving IP range [172.16.0.2, 172.16.0.2]
May  4 2018 19:03:58.910Z INFO  Reserving IP range [172.16.0.19, 172.16.0.19]
May  4 2018 19:03:58.999Z INFO  Reserving IP range [172.16.0.29, 172.16.0.29]
May  4 2018 19:03:59.099Z INFO  Reserving IP range [172.16.0.4, 172.16.0.4]
May  4 2018 19:03:59.158Z INFO  Reserving IP range [172.16.0.7, 172.16.0.7]
May  4 2018 19:03:59.215Z INFO  Reserving IP range [172.16.0.20, 172.16.0.20]
May  4 2018 19:03:59.259Z INFO  Reserving IP range [172.16.0.10, 172.16.0.10]
May  4 2018 19:03:59.330Z INFO  Reserving IP range [172.16.0.15, 172.16.0.15]
May  4 2018 19:03:59.421Z INFO  Reserving IP range [172.16.0.17, 172.16.0.17]
May  4 2018 19:03:59.490Z ERROR Cannot reserve IP range 172.16.0.38 - 172.16.0.38.  Already in use
May  4 2018 19:03:59.491Z FATAL could not initialize port layer: Cannot reserve IP range 172.16.0.38 - 172.16.0.38.  Already in use
time="2018-05-04T19:04:02Z" level=info msg="Launching portlayer server pprof server on 127.0.0.1:6063" 
time="2018-05-04T19:04:02Z" level=info msg="{Formatter:0xc42013bde0 Level:info Syslog:<nil>}" 
May  4 2018 19:04:03.041Z FATAL configure_port_layer ERROR: Failed to connect to wdc-stvc3944-146-dhcp15.eng.vmware.com: 503 Service Unavailable
```

I suspect that this is #7128 but not yet confirmed that one of the containers assigned `172.16.0.38` is powered off.

Digging through tether.debugs I've identified the following containers that have reported using this address:
a. jovial_lewin-8ae15c8ff431
b. confident_wilson-cca26ffb363f
c. nostalgic_visvesvaraya-9e6887c1053a
d. relaxed_goldwasser-63288eb62379
These belong to a VCH named `U-VIC-00023` and all are powered on with that address assigned.

There are three items of note here:
1. we should not be getting into this state - how did we end up with these cVMs with the same IP assigned?
2. we should not be restart the port layer in this case - it won't fix anything as it's a flawed configuration at fault. I suspect we're doing this due to not having an effective means of reporting a flawed config.
3. if restarting the portlayer we should _still_ be politely signing out of VC (which 6943 may accomplish if we stop using log.Fatal in this path)

I've shut down all of the VCHs that were hitting this issue as it's pointless just hammering VC with these repeated connection attempts.

---

@mdubya66 commented on [Mon May 07 2018](https://github.com/vmware/vic-product/issues/1676#issuecomment-387152221)

moving this to vic-engine

**Comment by mdubya66** [comment]
Might be related https://github.com/vmware/vic/issues/6851

**Comment by zjs** [comment]
Before closing this, it seems like we should re-test this case (concurrent upgrade of 50 VCHs) to verify that sessions are no longer leaking and that the upgrades succeed.

**Comment by renmaosheng** [comment]
@lgayatri could you please help on this? thanks.

**Comment by lgayatri** [comment]
We already verified this path for VIC 1.4,3 upgrade testing and had not seen the issue.
